### PR TITLE
Allow to generate stake addresses

### PR DESCRIPTION
**Pull request by hasufell**
Based on https://github.com/input-output-hk/cardano-addresses/pull/56

*TODO:*

- [x] tests
- [x] hlint
- [x] stylish

Built with:

- [x] cabal
- [x] stack
- [x] nix

### CODE REVIEW

File: core/lib/Cardano/Address/Style/Shelley.hs [hunk 1]
```diff
@@ -374,6 +376,19 @@ deriveStakingPrivateKey =
...
+    rewardAccAddress discrimination k = unsafeMkAddress $
+        invariantSize expectedLength $ BL.toStrict $ runPut $ do
+            putWord8 firstByte
+            putByteString (blake2b224 k)
+      where
+          -- First 4 bits are `1110` for keyhash28 reward account address.
+          -- Next 4 bits are network discriminator.
+          -- `1110 0000` is 224 in decimal.
+          firstByte =
+              224 + invariantNetworkTag 16 (networkTag @Shelley discrimination)
```
Comment (hasufell) [inline core/lib/Cardano/Address/Style/Shelley.hs:389]: This is the jist of it... I can verify this "works", but I'm not sure the semantics are correct.

File: command-line/lib/Command/Address/Reward.hs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+import qualified Data.ByteString.Char8 as B8
+import qualified Data.Text.Encoding as T
+
+
+newtype Cmd = Cmd
+    { networkTag :: NetworkTag
+    } deriving (Show)
+
+mod :: (Cmd -> parent) -> Mod CommandFields parent
+mod liftCmd = command "reward" $
```
Comment (KtorZ) [inline command-line/lib/Command/Address/Reward.hs:41]: Let's call this `stake` since this is the terminology used everywhere else.

File: command-line/lib/Command/Address/Reward.hs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    info (helper <*> fmap liftCmd parser) $ mempty
+        <> progDesc "Create a reward account address"
+        <> header "Create a reward account address \
+            \that references a staking key (1-1)."
+        <> footerDoc (Just $ vsep
+            [ string "The is read from stdin."
```
Comment (KtorZ) [inline command-line/lib/Command/Address/Reward.hs:47]: missing a word here :point_up:

File: command-line/lib/Command/Address/Reward.hs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+            , indent 2 $ bold $ string $ "$ "<>progName<>" recovery-phrase generate --size 15 \\"
+            , indent 4 $ bold $ string $ "| "<>progName<>" key from-recovery-phrase Shelley > root.prv"
+            , indent 2 $ string ""
+            , indent 2 $ bold $ string "$ cat root.prv \\"
+            , indent 4 $ bold $ string $ "| "<>progName<>" key child 1852H/1815H/0H/2/0 > stake.prv"
+            , indent 2 $ string ""
+            , indent 2 $ bold $ string "$ cat stake.prv \\"
+            , indent 4 $ bold $ string $ "| "<>progName<>" key public \\"
+            , indent 4 $ bold $ string $ "| "<>progName<>" address reward --network-tag 0"
+            , indent 2 $ string "addr1uqly0fjvrgguywze067gwhsexggtj8rrdnxczgp5vexe8zgxqns3g"
```
Comment (KtorZ) [inline command-line/lib/Command/Address/Reward.hs:59]: :+1:

File: core/lib/Cardano/Address.hs [hunk 5]
```diff
@@ -121,6 +122,16 @@ fromBech32 :: Text -> Maybe Address
...
 
+-- | Encoding of addresses for certain key types and backend targets.
+--
+-- @since 1.0.0
+class HasNetworkDiscriminant key => RewardAccAddress key where
+    -- | Convert a staking key to a reward account 'Address' valid for the given
+    -- network discrimination.
+    --
+    -- @since 1.0.0
+    rewardAccAddress :: NetworkDiscriminant key -> key 'StakingK XPub -> Address
```
[the remark this answers is not in the document]
Comment (hasufell) [reply core/lib/Cardano/Address.hs:133]: I'm confused enough about all this. We probably need a catchup call about all the addresses.

I tried to stick to the terminology from the shelley design spec. Doesn't that say "Reward Account Address"?

File: core/lib/Cardano/Address/Style/Shelley.hs (hunk 1 above)
Comment (hasufell) [reply core/lib/Cardano/Address/Style/Shelley.hs:391]: @jonathanknowles I know this is a magic number... but I really cannot say what name would be reasonable here.

Comment (KtorZ) [reply core/lib/Cardano/Address/Style/Shelley.hs:391]: `addrType` would be a reasonable name if you're looking for inspiration.

File: core/lib/Cardano/Address.hs [hunk 6]
```diff
@@ -121,6 +122,16 @@ fromBech32 :: Text -> Maybe Address
 fromBech32 =
     eitherToMaybe . fmap unsafeMkAddress. E.fromBech32 (const id) . T.encodeUtf8
 
+-- | Encoding of addresses for certain key types and backend targets.
+--
+-- @since 1.0.0
```
Comment (KtorZ) [inline core/lib/Cardano/Address.hs:127]: *2.0.0

File: command-line/lib/Command/Address/Reward.hs [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        <$> networkTagOpt Shelley
+
+run :: Cmd -> IO ()
+run Cmd{networkTag} = do
+    xpub <- hGetXPub stdin
+    case (mkNetworkDiscriminant . fromIntegral . unNetworkTag) networkTag of
+        Left e -> die (fmt $ build e)
+        Right discriminant -> do
+            let addr = Shelley.stakeAddress discriminant (Shelley.liftXPub xpub)
+            B8.hPutStr stdout $ T.encodeUtf8 $ bech32 addr
```
Comment (KtorZ) [inline command-line/lib/Command/Address/Reward.hs:76]: This should use [`bech32With`](https://input-output-hk.github.io/cardano-addresses/haddock/cardano-addresses-1.0.0/Cardano-Address.html#v:bech32With) and provide an appropriate prefix:

- `stake` for mainnet 
- `stake_test` for testnet

File: core/lib/Cardano/Address/Style/Shelley.hs (hunk 1 above)
Comment (hasufell) [reply core/lib/Cardano/Address/Style/Shelley.hs:391]: While at it... what's the magic number 16?

File: command-line/lib/Command/Address/Reward.hs (hunk 7 above)
Comment (hasufell) [reply command-line/lib/Command/Address/Reward.hs:76]: And what about values outside of 0 and 1?

Comment (hasufell) [reply command-line/lib/Command/Address/Reward.hs]: I just use `addr` for these.

Comment (KtorZ) [reply command-line/lib/Command/Address/Reward.hs]: > And what about values outside of 0 and 1?

Right now they should be `stake`. `0` is testnet, and `1` designates the _current_ mainnet. Values beyond 1 are reserved for future mainnets. There are none currently.

File: command-line/lib/Command/Address/Reward.hs [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+run Cmd{networkTag} = do
+    xpub <- hGetXPub stdin
+    hpr <- hprFromNetTag
+    case (mkNetworkDiscriminant . fromIntegral . unNetworkTag) networkTag of
+        Left e -> die (fmt $ build e)
+        Right discriminant -> do
+            let addr = Shelley.stakeAddress discriminant (Shelley.liftXPub xpub)
+            B8.hPutStr stdout $ T.encodeUtf8 $ bech32With hpr addr
+  where
+    hprFromText = either throwM pure . humanReadablePartFromText
```
Comment (KtorZ) [inline command-line/lib/Command/Address/Reward.hs:84]: NOTE: you can use `bech32-th` + QuasiQuotes for literal HRP in such fashion:

```hs
| shelleyMainnet == networkTag =  [humanReadablePart|stake|]
| shelleyTestnet == networkTag = [humanReadablePart|stake_test|]
```

File: command-line/lib/Command/Address/Reward.hs [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+    hprFromNetTag
+        | shelleyMainnet == networkTag = hprFromText "stake"
+        | shelleyTestnet == networkTag = hprFromText "stake_test"
+        | otherwise = hprFromText "addr"
```
Comment (KtorZ) [inline command-line/lib/Command/Address/Reward.hs:88]: `stake` would be a better default for values beyond `1`. Or, alternatively, we could change the current parser for the network tag option to fail if provided anything different than `0` or `1` and use instead named constant like `testnet` & `mainnet` (nicely done in a separate PR).

**Review by KtorZ — APPROVED** [review]

File: command-line/test/Command/Address/RewardSpec.hs [hunk 10]
```diff
@@ -17,7 +17,7 @@ spec = describeCmd [ "address", "stake" ] $ do
         "stake_test1ura3dk68y6echdmfmnvm8mej8u5truwv8ufmv830w5a45tcsfhtt2"
 
     specShelley defaultPhrase "1852H/1815H/0H/2/0" 3
-        "addr1u0a3dk68y6echdmfmnvm8mej8u5truwv8ufmv830w5a45tc3kzy0t"
+        "stake1u0a3dk68y6echdmfmnvm8mej8u5truwv8ufmv830w5a45tchw5z0e"
 
     specMalformedNetwork "💩"
```
Comment (hasufell) [inline command-line/test/Command/Address/RewardSpec.hs:22]: I love this one :laughing:

Comment (KtorZ) [reply command-line/test/Command/Address/RewardSpec.hs:22]: :) It is very important. In particular, this one caught some nice regressions on Windows Powershell.
